src/Primitives/src/StringSegment.cs

```diff
@@ -564,67 +498,78 @@ public int IndexOfAny(char[] anyOf)
         /// <returns>The zero-based index position of value if that character is found, or -1 if it is not.</returns>
         public int LastIndexOf(char value)
         {
-            if (!HasValue)
-            {
-                return -1;
-            }
+            var index = -1;
 
-            var index = Buffer.LastIndexOf(value, Offset + Length - 1, Length);
-            if (index == -1)
+            if (HasValue)
             {
-                return -1;
+                index = Buffer.LastIndexOf(value, Offset + Length - 1, Length);
+                if (index != -1)
+                {
+                    index -= Offset;
+                }
             }
 
-            return index - Offset;
+            return index;
         }
 
         /// <summary>
         /// Removes all leading and trailing whitespaces.
         /// </summary>
         /// <returns>The trimmed <see cref="StringSegment"/>.</returns>
-        public StringSegment Trim()
-        {
-            return TrimStart().TrimEnd();
-        }
+        public StringSegment Trim() => TrimStart().TrimEnd();
 
         /// <summary>
         /// Removes all leading whitespaces.
         /// </summary>
         /// <returns>The trimmed <see cref="StringSegment"/>.</returns>
-        public StringSegment TrimStart()
+        public unsafe StringSegment TrimStart()
         {
             var trimmedStart = Offset;
-            while (trimmedStart < Offset + Length)
+            var length = Offset + Length;
+
+            fixed (char* p = Buffer)
```

Comment:
Is this to get rid of the bounds checks? Is there no other clever trick that would remove the need for unsafe code?